### DISCUSSION TITLE
Proposal: mechanism to exclude custom `record` fields from `with` cloning

### DISCUSSION THREAD

**Discussion by masonwheeler** [dotnet/csharplang#6133] [Language Ideas]
Currently, when cloning a `record` with a `with` expression, all fields of the record are copied, including any custom fields that may have been added by the developer.  This is generally a good idea, but there are times when it can yield undesirable behavior.  (For example, it's quite plausible that a `record` could have a custom field containing data computed from primary `record` properties and cached, and then the `with` operation changes the primary properties that the cache is based on, leaving invalid, stale data cached in the clone.)

There is currently no mechanism to designate a field as non-cloneable.  [The documentation](https://docs.microsoft.com/en-us/dotnet/csharp/language-reference/operators/with-expression#custom-copy-semantics) gives the rather bleak advice that if you want any non-default cloning semantics, you need to write your own custom version of the copy constructor.  Unfortunately this means that any future changes in the `record` declaration need to be manually kept in sync with the custom copy constructor, which somewhat defeats the purpose of having the compiler's `record` processing set this stuff up for you in the first place.

It would be nice to instead be able to annotate such a field with an attribute that tells the `record` code not to clone it.  For example:

```
   public record MyData(string Name, DataElement[] Data)
   {
      [NoWithClone]
      private int _cost;

      public int Cost 
      {
         if (_cost == 0)
            _cost = Data.Select(d => CostUtils.ExpensiveComputeOperation(d)).Sum();
         return _cost;
      }
   }
```

If you use `with` to change the `Data` on this record, a previously-computed `Cost` becomes invalid.  Therefore, its backing field is annotated with a `[NoWithClone]` attribute to ensure that it will not be copied to the new record.  (The name is admittedly a bit awkward.  I'd be just fine with someone coming up with a better name.)

**Comment by RikkiGibson** [comment]
> Unfortunately this means that any future changes in the record declaration need to be manually kept in sync with the custom copy constructor, which somewhat defeats the purpose of having the compiler's record processing set this stuff up for you in the first place.

If this is a concern, I think the advice would be to actually invent a `[NoWithClone]` attribute yourself, and a source generator that recognizes it, and have *it* generate your copy constructor.

**Reply by masonwheeler** [reply]
"If you don't like it, build your own" might make sense if I was the only one having this problem, but I'm not.  [This is currently the third-highest-ranked topic on StackOverflow's "c#-record-type" tag,](https://stackoverflow.com/questions/tagged/c%23-record-type?tab=Votes) which would seem to indicate that it's a legitimate problem for people using records.

**Reply by CyrusNajmabadi** [reply]
It may be something that people can use assistance on.  That doesn't mean a language feature is the preferred way to address it.  It may be the case that an attribute+SG is the way to go.  That would not be in csharplang, but likely something more like the sdk.

**Reply by jnm2** [reply]
It's come up for me a few times now, and it's a pit of failure as a pattern that is common in ordinary classes that you don't tend to think much about.

**Reply by HaloFour** [reply]
Spaghetti at the wall, but maybe the compiler can emit partial method declarations within the record giving the developer some known interception points at which they can interject logic?  Then instead of having to replace a member outright, they can just implement those partial methods.

```cs
public record MyData(string Name, DataElement[] Data)
   {
      private int _cost;

      public int Cost 
      {
         if (_cost == 0)
            _cost = Data.Select(d => CostUtils.ExpensiveComputeOperation(d)).Sum();
         return _cost;
      }

      // called at the end of the copy constructor to allow custom fields to be reset
      partial void AfterClone()
      {
        _cost = 0;
      }
   }
```

**Reply by masonwheeler** [reply]
Yeah, some defined hook point was actually the first thing I looked for before looking for a way to suppress cloning.  Unfortunately, neither of them exists in the current implementation.

**Reply by CyrusNajmabadi** [reply]
I think from our last convo on the topic, we felt the extensino point was to write any of these methods yourself (we allow providing an impl for nearly all of them).  If the boilerplate nature of that isn't pleasant, that's exactly what an SG is good for. 

I think we'd need a strong example of the SG approach not being sufficient to warrant a lang change here.  Note: the SG approach doesn't mean *you'd* need to write it yourself, just that it be provided by some group.  So you'd have a solution, it just wouldn't be baked into the language, it would be from an ancillary team.

**Comment by HaloFour** [comment]
The team intended on revisiting "withers" to make them pattern-based so that any type could support them.  If that leads to being able to support custom "withers" (which would likely start with a clone method) that sounds like a good place to offer this customization.

**Comment by Clockwork-Muse** [comment]
Alternatively - wrap the source and cache in a type separate from the rest of the variables.

**Reply by jnm2** [reply]
This doesn't work well when a property value is calculated and cached from two other properties, for instance. Adding levels of nesting is also fairly annoying when using today's `with` which doesn't support dotted properties.

**Reply by masonwheeler** [reply]
Agreed. Plus, creating new types everywhere for every little detail is bad aesthetics and bad for maintenance.  I'd really prefer not to turn my C# code into the mess of ravioli that has gotten Java enterprise code laughed at for decades.

**Comment by TonyValenti** [comment]
I would prefer to be able to have a method named "with" that gets called after the default "with" is called that lets me fix up any cached/computed properties.

**Reply by HaloFour** [reply]
The problem with that route is that there is no "end" to a with expression.  It's a "clone" followed by any number of writes to `init` or `set` accessors.  This would require some kind of final method to be called to indicate when the "with" is complete, which downlevel compilers would ignore since they would not be aware of it.

Other than that I don't think it's a bad idea to have a way to indicate that the "initializer" of a type has completed, wither or not.

**Reply by masonwheeler** [reply]
What do you mean by "downlevel compilers"?  I'm not familiar with the term.

**Reply by HaloFour** [reply]
> What do you mean by "downlevel compilers"? I'm not familiar with the term.

Older versions of the compiler.
